File: Haskell/2024/02.hs
import Lib

type Input = [[Int]]

main :: IO ()
main = aoc 2024 2 setup solve1 solve2 ["1"]

solve1 :: Input -> Int
solve1 = length . filter check

solve2 :: Input -> Int
solve2 = length . filter (any check . subLists)

setup :: String -> Input
setup = map (map read . words) . lines

check :: [Int] -> Bool
check = fork (&&) checkMonotone checkRange . diffs

checkMonotone :: [Int] -> Bool
checkMonotone = (((==) . head) >>= all) . map (< 0)

checkRange :: [Int] -> Bool
checkRange = all ((`elem` [1 .. 3]) . abs)

diffs :: [Int] -> [Int]
diffs = zipWith (-) <*> tail

subLists :: [a] -> [[a]]
subLists = fork map (flip removeAt) indices

indices :: [a] -> [Int]
indices = flip take [0 ..] . length

removeAt :: Int -> [a] -> [a]
removeAt = fork (fork (++)) take (drop . (+ 1))

File: Ruby/2024/02.rb
def check(nums)
  d = nums[...-1].zip(nums[1..]).map { _1.reduce &:- }
  return d.all? { (1..3).include?(_1.abs) && _1.negative? == d[0].negative? }
end

input = File.read(ARGV[0]).lines.map { _1.split.map &:to_i }

p input.filter { check _1 }.length
p input.filter { |nums| (0...nums.length).any? { check nums[..._1] + nums[_1+1..] } }.length
